File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/ARID.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bcrand.randomData
import com.blockchaincommons.bctags.TAG_ARID
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * An "Apparently Random Identifier" (ARID).
 *
 * An ARID is a cryptographically strong, universally unique identifier with
 * the following properties:
 * - Non-correlatability: The bit sequence cannot be correlated with its
 *   referent or any other ARID
 * - Neutral semantics: Contains no inherent type information
 * - Open generation: Any method of generation is allowed as long as it
 *   produces statistically random bits
 * - Minimum strength: Must be 256 bits (32 bytes) in length
 * - Cryptographic suitability: Can be used as inputs to cryptographic
 *   constructs
 *
 * Unlike digests/hashes which identify a fixed, immutable state of data,
 * ARIDs can serve as stable identifiers for mutable data structures.
 *
 * As defined in
 * [BCR-2022-002](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2022-002-arid.md).
 */
class ARID private constructor(private val data: ByteArray) :
    ReferenceProvider,
    Comparable<ARID>,
    CborTaggedCodable,
    URCodable {

    init {
        require(data.size == ARID_SIZE) {
            "ARID data must be exactly $ARID_SIZE bytes, got ${data.size}"
        }
    }

    /** Returns a copy of the underlying 32-byte ARID data. */
    fun data(): ByteArray = data.copyOf()

    /** Returns the ARID bytes as a copy. */
    fun asBytes(): ByteArray = data.copyOf()

    /** The ARID as a 64-character lowercase hexadecimal string. */
    val hex: String get() = data.toHexString()

    /** Returns the first four bytes of the ARID as a hexadecimal string. */
    fun shortDescription(): String = data.copyOfRange(0, 4).toHexString()

    // -- ReferenceProvider --

    override fun reference(): Reference =
        Reference.fromDigest(Digest.fromImage(taggedCbor().toCborData()))

    // -- Comparable --

    override fun compareTo(other: ARID): Int {
        for (i in data.indices) {
            val cmp = (data[i].toInt() and 0xFF) - (other.data[i].toInt() and 0xFF)
            if (cmp != 0) return cmp
        }
        return 0
    }

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ARID) return false
        return data.contentEquals(other.data)
    }

    override fun hashCode(): Int = data.contentHashCode()

    // -- toString --

    override fun toString(): String = "ARID($hex)"

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_ARID))

    override fun untaggedCbor(): Cbor = Cbor.fromByteString(data)

    companion object {
        const val ARID_SIZE: Int = 32

        /** Creates a new random ARID. */
        fun create(): ARID {
            val data = randomData(ARID_SIZE)
            return ARID(data)
        }

        /** Restores an ARID from exactly [ARID_SIZE] bytes. */
        fun fromData(data: ByteArray): ARID {
            if (data.size != ARID_SIZE) {
                throw BcComponentsException.invalidSize("ARID", ARID_SIZE, data.size)
            }
            return ARID(data.copyOf())
        }

        /**
         * Creates an ARID from a hexadecimal string.
         *
         * @throws IllegalArgumentException if the string is not exactly 64
         *   hex digits.
         */
        fun fromHex(hex: String): ARID = fromData(hex.hexToByteArray())

        /** Decodes an [ARID] from untagged CBOR (a byte string). */
        fun fromUntaggedCbor(cbor: Cbor): ARID {
            val bytes = cbor.tryByteStringData()
            return fromData(bytes)
        }

        /** Decodes an [ARID] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): ARID =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_ARID)),
            ) { fromUntaggedCbor(it) }

        /** Decodes an [ARID] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): ARID =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_ARID)),
            ) { fromUntaggedCbor(it) }

        /** Decodes an [ARID] from a UR. */
        fun fromUr(ur: UR): ARID {
            ur.checkType("arid")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes an [ARID] from a UR string. */
        fun fromUrString(urString: String): ARID =
            fromUr(UR.fromUrString(urString))
    }
}

File: csharp/BCComponents/BCComponents/ARID.cs
using BlockchainCommons.BCRand;
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// An "Apparently Random Identifier" (ARID).
/// </summary>
/// <remarks>
/// <para>
/// An ARID is a cryptographically strong, universally unique identifier with
/// the following properties:
/// </para>
/// <list type="bullet">
/// <item>Non-correlatability: The bit sequence cannot be correlated with its
/// referent or any other ARID</item>
/// <item>Neutral semantics: Contains no inherent type information</item>
/// <item>Open generation: Any method of generation is allowed as long as it
/// produces statistically random bits</item>
/// <item>Minimum strength: Must be 256 bits (32 bytes) in length</item>
/// <item>Cryptographic suitability: Can be used as inputs to cryptographic
/// constructs</item>
/// </list>
/// <para>
/// Unlike digests/hashes which identify a fixed, immutable state of data,
/// ARIDs can serve as stable identifiers for mutable data structures.
/// </para>
/// </remarks>
public sealed class ARID
    : IEquatable<ARID>,
      IComparable<ARID>,
      IReferenceProvider,
      ICborTaggedEncodable,
      ICborTaggedDecodable
{
    /// <summary>The size of an ARID in bytes.</summary>
    public const int Size = 32;

    private readonly byte[] _data;

    private ARID(byte[] data)
    {
        _data = data;
    }

    /// <summary>Creates a new random ARID.</summary>
    /// <returns>A new <see cref="ARID"/>.</returns>
    public static ARID New()
    {
        return NewUsing(SecureRandomNumberGenerator.Shared);
    }

    /// <summary>Creates a new random ARID using the given RNG.</summary>
    /// <param name="rng">The random number generator to use.</param>
    /// <returns>A new <see cref="ARID"/>.</returns>
    public static ARID NewUsing(IRandomNumberGenerator rng)
    {
        return new ARID(rng.RandomData(Size));
    }

    /// <summary>
    /// Creates an ARID from exactly <see cref="Size"/> bytes.
    /// </summary>
    /// <param name="data">Exactly 32 bytes of ARID data.</param>
    /// <returns>A new <see cref="ARID"/>.</returns>
    /// <exception cref="BCComponentsException">
    /// Thrown if <paramref name="data"/> is not exactly 32 bytes.
    /// </exception>
    public static ARID FromData(byte[] data)
    {
        if (data.Length != Size)
            throw BCComponentsException.InvalidSize("ARID", Size, data.Length);
        return new ARID((byte[])data.Clone());
    }

    /// <summary>Creates an ARID from a hexadecimal string.</summary>
    /// <param name="hex">A 64-character hexadecimal string.</param>
    /// <returns>A new <see cref="ARID"/>.</returns>
    public static ARID FromHex(string hex)
    {
        return FromData(Convert.FromHexString(hex));
    }

    /// <summary>Returns a copy of the underlying 32-byte ARID data.</summary>
    public byte[] Data => (byte[])_data.Clone();

    /// <summary>Returns the ARID bytes as a copy.</summary>
    public byte[] AsBytes() => (byte[])_data.Clone();

    /// <summary>Gets the ARID as a 64-character lowercase hexadecimal string.</summary>
    public string Hex => Convert.ToHexString(_data).ToLowerInvariant();

    /// <summary>Returns the first four bytes of the ARID as a hexadecimal string.</summary>
    public string ShortDescription() =>
        Convert.ToHexString(_data, 0, 4).ToLowerInvariant();

    // --- IReferenceProvider ---

    /// <inheritdoc/>
    public Reference Reference() =>
        BCComponents.Reference.FromDigest(Digest.FromImage(TaggedCbor().ToCborData()));

    // --- IComparable<ARID> ---

    /// <inheritdoc/>
    public int CompareTo(ARID? other)
    {
        if (other is null) return 1;
        return _data.AsSpan().SequenceCompareTo(other._data);
    }

    // --- IEquatable<ARID> ---

    /// <inheritdoc/>
    public bool Equals(ARID? other)
    {
        if (other is null) return false;
        return _data.AsSpan().SequenceEqual(other._data);
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is ARID a && Equals(a);

    /// <inheritdoc/>
    public override int GetHashCode()
    {
        var hash = new HashCode();
        foreach (var b in _data)
            hash.Add(b);
        return hash.ToHashCode();
    }

    /// <summary>Tests equality of two ARID instances.</summary>
    public static bool operator ==(ARID? left, ARID? right) =>
        left is null ? right is null : left.Equals(right);

    /// <summary>Tests inequality of two ARID instances.</summary>
    public static bool operator !=(ARID? left, ARID? right) => !(left == right);

    // --- ICborTagged ---

    /// <summary>Returns the CBOR tags for this type (40012).</summary>
    public static IReadOnlyList<Tag> CborTags =>
        GlobalTags.TagsForValues(BcTags.TagArid);

    // --- ICborTaggedEncodable ---

    /// <summary>Returns the untagged CBOR representation (a byte string).</summary>
    public Cbor UntaggedCbor() => Cbor.ToByteString(_data);

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor ToCbor() => TaggedCbor();

    // --- ICborTaggedDecodable ---

    /// <summary>Decodes an <see cref="ARID"/> from untagged CBOR (a byte string).</summary>
    /// <param name="cbor">The untagged CBOR value.</param>
    /// <returns>A new <see cref="ARID"/>.</returns>
    public static ARID FromUntaggedCbor(Cbor cbor)
    {
        var data = cbor.TryIntoByteString();
        return FromData(data);
    }

    /// <summary>Decodes an <see cref="ARID"/> from tagged CBOR.</summary>
    /// <param name="cbor">The tagged CBOR value.</param>
    /// <returns>A new <see cref="ARID"/>.</returns>
    public static ARID FromTaggedCbor(Cbor cbor)
    {
        foreach (var tag in CborTags)
        {
            try
            {
                var item = cbor.TryIntoExpectedTaggedValue(tag);
                return FromUntaggedCbor(item);
            }
            catch (CborWrongTagException) { }
            catch (CborWrongTypeException) { }
        }
        throw new CborWrongTypeException();
    }

    // --- Display ---

    /// <inheritdoc/>
    public override string ToString() => $"ARID({Hex})";
}
